sample-app/src/components/DropdownSection.tsx

```diff
@@ -0,0 +1,108 @@
+import classNames from "classnames";
+import { useState, useEffect } from "react";
+
+export interface Option {
+  value: string,
+  display: JSX.Element
+}
+
+interface DropdownSectionProps<T> {
+  focusStatus?: string,
+  options: T[],
+  optionIdPrefix: string,
+  onFocusChange: (value: string, focusedOptionId: string) => void,
+  onLeaveSectionFocus?: (focusNext: boolean) => void,
+  onClickOption: (option: T, optionIndex: number) => void,
+  label?: string,
+  cssClasses: {
+    sectionContainer: string,
+    sectionLabel: string,
+    optionsContainer: string,
+    option: string,
+    focusedOption: string
+  }
+}
+
+export default function DropdownSection<T extends Option>({
+  focusStatus,
+  options,
+  optionIdPrefix,
+  onFocusChange,
+  onLeaveSectionFocus = () => {},
+  onClickOption,
+  label = '',
+  cssClasses
+}: DropdownSectionProps<T>): JSX.Element | null {
+
+  const [focusedOptionIndex, setFocusedOptionIndex] = useState<number>(0);
+
+  function changeOptionFocus(focusNext: boolean) {
+    let newIndex = focusNext ? focusedOptionIndex + 1 : focusedOptionIndex - 1;
+
+    if (newIndex > -1 && newIndex < options.length) {
+      onFocusChange(options[newIndex].value, `${optionIdPrefix}-${newIndex}`);
+    } else {
+      focusNext ? onLeaveSectionFocus(true) : onLeaveSectionFocus(false);
+      newIndex = newIndex === -1 ? 0 : options.length - 1;
+    }
+
+    setFocusedOptionIndex(newIndex);
+  }
+
+  function handleKeyDown(evt: globalThis.KeyboardEvent) {
+    if (focusStatus === 'active') {
+      if (['ArrowDown', 'ArrowUp', 'ArrowRight', 'ArrowLeft'].includes(evt.key)) {
+        evt.preventDefault();
+      }
+
+      if (evt.key === 'ArrowDown' || evt.key === 'ArrowRight') {
+        changeOptionFocus(true);
+      } else if (evt.key === 'ArrowUp' || evt.key === 'ArrowLeft') {
+        changeOptionFocus(false);
+      }
+    }
+  }
+
+  useEffect(() => {
+    if (focusStatus === 'active') {
+      onFocusChange(options[focusedOptionIndex].value, `${optionIdPrefix}-${focusedOptionIndex}`);
+    } else if (focusStatus === 'reset') {
```

Comment:
can we get rid of 'reset' status? from what I can see, it's use when focusedSectionIndex is undefined, which is when everything is hidden anyway. Not much different from 'inactive' status

Reply:
the 'reset' status is how the sections know to reset their `focusedOptionIndex` to the beginning of the section. otherwise, it could result in incorrect option focus behavior after a search is performed. because the section would remember the last place it was focused as the default

Reply:
I think during focus shift, once a section is marked with a focus of 'inactive' then it can reset to the beginning of the section then.

Reply:
right now, the way a section is maintaining knowledge of which side it gains focus from (start or end) depends on what the focused option index is when 'inactive'. if this were reset, when a section becomes active, it wouldn't know if it should focus on the last option or the first

Reply:
updated to remove 'reset' status by changing the key of the child component instead